ci: proposal - update Jest Snaps at release (#1300)
coveo/cli: rfcs/accepted/0002-cd-update-snap.md

- Feature Name: Update Jest Snap during release process
- Start Date: 2023-04-25

# Summary

[summary]: #summary

> Move and reworks some release steps to update Jest Snap when we do a release of the CLI while not doubling the release time.

Reviewer: ```suggestion
Move and rework some release steps to update Jest Snap when we do a release of the CLI while not doubling the release time.
```

# Motivation

[motivation]: #motivation

> When a release of the CLI happens, it will now makes some Jest Snap outdated, which causes the tests to fail.

Reviewer: ```suggestion
When a release of the CLI happens, it currently makes some Jest Snaps outdated causing the tests to fail.
```

# Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

Our current release process looks like this

```mermaid
graph LR
    A(npm) --> B(Git) --> C(GitHub) --> D(Build)
```

1. We release to NPM
2. We create and push a release commit and associated git tags
3. We create a GitHub release for the CLI (if the CLI has changed)
4. We build the CLI and add the binaries to the GitHub release

We need the release commit to include the snap udpates to keep the main branch green.
To do so while not hampering our release process too much, our release process would change to look like this:

```mermaid
graph LR
    A(npm)  --> B(E2E)      --> C(Git) --> D(GitHub)
    A       --> B2(Build)   --> D
```

1. We release to NPM
2. The release process split into two branches:
   - One branch start building the binaries right away.
     - It stores them in a GitHub Job Artifact.

>      - It 'noop' if the CLI does not to be build.

Reviewer: Adding some info on when the CLI would not need to be built would be helpful

>      - It 'noop' if the CLI does not to be build.

Reviewer: ```suggestion
     - It would be a 'noop' if the CLI does not need to be build.
```

>    - The other branch refresh the Jest Snap and then proceed with the previous release process up to the Git commit/push.

Reviewer: ```suggestion
   - The other branch refreshes the Jest Snap and then proceeds with the release process as before up to the Git commit/push.
```

3. The creation of the GitHub release now occurs when both Git and the binary jobs are done, and only if the CLI needs to be built/released.

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

- Split the Release Workflow into several workflows, one by step
- NPM job would export a git patch so that E2E and Build can continue without any hitches.
- E2E and Git can be one job split into two steps

# Drawbacks

[drawbacks]: #drawbacks

- Potential increase in complexity

> - Lack of control on our assertions: a bot will automatically update some assertions

Reviewer: what are the "assertions" referring to here?

Author: ```suggestion
- Lack of control over our assertions: a bot will automatically update the Jest Snaps, which are what we 'expect' the code to do. (e.g `expect(something).toMatchSnapshot()`)
```

# Rationale and alternatives

[rationale-and-alternatives]: #rationale-and-alternatives

- Do nothing: a maintainer will have to update the snap after each release
- Create automatically a PR to update the Jest Snap:
  - Pro: we have more control on our assertions
  - Cons:
    - Daily E2E will fail
    - We lose the 'always green' status of the main branch

# Prior art

[prior-art]: #prior-art

> It's really and oddity, I didn't find much prior art.

Reviewer: ```suggestion
It's really an oddity, I didn't find much prior art.
```
